### ISSUE TITLE
GlyphLayout with Align.center should respect BitmapFont.usesIntegerPositions()

### ISSUE THREAD

**Issue by dan-fitzgerald-av** [libgdx/libgdx#3254]
I'm running into an issue with nearest-neighbor filtered BitmapFonts with integer positioning, center alignment, and multi-line wrapping:
![java 2015-07-02 13-17-35-49](https://cloud.githubusercontent.com/assets/3443800/8484609/301b3abe-20be-11e5-83eb-05185cdc83e4.png)
![java 2015-07-02 13-17-41-73](https://cloud.githubusercontent.com/assets/3443800/8484610/3185bffa-20be-11e5-8f59-ac2e20f0fbdb.png)

I think it could be solved around lines 210 and 218 of GlyphLayout (looking at 1.6.3) by rounding the `shift` variable in both places if `font.usesIntegerPositions()` is true. Not sure, though!

**Comment by xoppa** [comment]
See https://github.com/libgdx/libgdx/blob/master/CONTRIBUTING.md. Please include the required information to reproduce the issue.

**Comment by dan-fitzgerald-av** [comment]
On an lines that aren't the widest line of a multi-line, center aligned BitmapFont, some of the glyphs get garbled as seen in the "M" of "TIME" in the first image or the "W" of "WHIP." 

There appears to be both vertical and horizontal issues in the line that say "143 Enemies" here:
![java 2015-07-02 14-02-23-80](https://cloud.githubusercontent.com/assets/3443800/8485629/989d4068-20c4-11e5-95f5-6f598cc1c9d4.png)

For reference, here's a still from the old version of the game, which used 1.5.5, before the big BitmapFont overhaul:
![cap1 avi_snapshot_07 20_ 2015 07 02_14 17 25](https://cloud.githubusercontent.com/assets/3443800/8486106/ff30d878-20c7-11e5-8c92-04d847241337.png)

Based on some tweaks I had just made to my own text wrapper class' alignment offset code (I was seeing similar issues in even single line text at first,) I'm fairly certain that the change I suggested to GlyphLayout will fix the issue. Here is the edit I would make (I added three lines):

```
		// line 200 in GlyphLayout as of 1.6.3
		// Align runs to center or right of targetWidth.
		if ((halign & Align.left) == 0) { // Not left aligned, so must be center or right aligned.
			boolean center = (halign & Align.center) != 0;
			boolean integer = font.usesIntegerPositions(); // Added this line
			float lineWidth = 0, lineY = Integer.MIN_VALUE;
			int lineStart = 0, n = runs.size;
			for (int i = 0; i < n; i++) {
				GlyphRun run = runs.get(i);
				if (run.y != lineY) {
					lineY = run.y;
					float shift = targetWidth - lineWidth;
					if (center) shift /= 2;
					if (integer) shift = Math.round(shift); // Added this line
					while (lineStart < i)
						runs.get(lineStart++).x += shift;
					lineWidth = 0;
				}
				lineWidth += run.width;
			}
			float shift = targetWidth - lineWidth;
			if (center) shift /= 2;
			if (integer) shift = Math.round(shift); // Added this line
			while (lineStart < n)
				runs.get(lineStart++).x += shift;
		}
```

I tried reproducing it with the built in arial-15 font but the issue wasn't apparent as it is with a low res pixel font like I'm using. Do I need to make and upload a test project that includes the .fnt and textures to reproduce? Should I be doing a pull request instead?

**Comment by NathanSweet** [comment]
The rounding happens as late as possible, in BitmapFontCache. The question
is how you could have seen the rounding not happen?

**Comment by dan-fitzgerald-av** [comment]
Thanks for the reply Nathan!

I think I figured it out, and I stand by my original suggestion (although I think I have a neater one now.)

When an x or y value that ends in a .5 is passed to the `addGlyph` method in BitmapFontCache, there's a good chance that the resulting x and x2 or y and y2 will round in opposite directions, causing issues like mine. This is why I had to start rounding my own offset values after updating to the new BitmapFont refactor, and why GlyphLayout's centering is causing me issues.

My new suggestion is to round `gx` and `gy` in the `addToCache` method in BitmapFontCache. There might still be rounding errors with the spacing between glyphs and/or runs, though, so maybe there should be rounding further up the chain!

Another solution might be to use floor instead of round, but I haven't fully thought that through.
